In [1]:
import pandas as pd

#load the data raw and normalized
Data_raw_and_normalized = pd.read_csv("../Data/TephraDataSet_raw_and_normalized.csv", encoding = 'latin1', low_memory =False )

 In the dataset the data was stored as in the original publication. Because of this, in many cases the assigned value to a particular element might not be a number but an expresion to indicate: missingness, lower than a certain value, below detection limit (bdl) or over range. In order to treat the data these values are replaced by NAN. In order to process the data it is necesary to replace these expressions by "NaN".

In [2]:
import numpy as np
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='Over range',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='bdl',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='<0.01', value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='<0.1',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='<1',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='<5',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='<6',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='<10',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='n.a.',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='Not analyzed',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='-',value=np.nan)
Data_raw_and_normalized = Data_raw_and_normalized.replace(to_replace='Not Netermined',value=np.nan)

For evaluating the effect that analyzing different major elements has on their normalization, we are only interested in Sample Observations where major elements have been analyzed:

In [3]:
Data_raw_and_normalized_major = Data_raw_and_normalized.dropna(subset=['SiO2'],axis=0,how='any')
#Data_raw_and_normalized_major = Data_raw_and_normalized_major.dropna(subset=['properties.MeasurementRun'],axis=0,how='any')

We sort values in order to better understand the distribution of the missingness of MnO, P2O5 and Cl:

In [4]:
Data_raw_and_normalized_major = Data_raw_and_normalized_major.replace(np.nan,-1)
temp1 = Data_raw_and_normalized_major[Data_raw_and_normalized_major.Total != -1]

temp1_1 = temp1[temp1.Cl != -1]
temp1_1_1 = temp1_1[temp1_1.P2O5 != -1]
case1 = temp1_1_1[temp1_1_1.MnO != -1]; case2 = temp1_1_1[temp1_1_1.MnO == -1]
print('\033[1mCase 1: Muestras con Cl, P2O5 y MnO: \033[0m {}'.format(len(case1)))
temp1_1_1_1_ms =  case1.replace(np.nan,-1)
print('From samples in Case 1, ' + str(len(temp1_1_1_1_ms[temp1_1_1_1_ms.MeasurementRun!=-1])) + ' have corresponding geostandard information analyzed along with the samples')
print('\033[1mCase 2: Muestras con Cl, con P2O5 y sin MnO: \033[0m {}'.format(len(case2)))
case1 = case1.replace(-1,np.nan);case2 = case2.replace(-1,np.nan)

temp1_1_2 = temp1_1[temp1_1.P2O5 == -1]
temp1_1_2_1 = temp1_1_2[temp1_1_2.MnO != -1];temp1_1_2_2 = temp1_1_2[temp1_1_2.MnO == -1]
case3 = temp1_1_2_1.replace(-1,np.nan); case4 = temp1_1_2_2.replace(-1,np.nan)
print('\033[1mCase 3: Muestras con Cl sin P2O5 y con MnO: \033[0m {}'.format(len(case3)))
print('\033[1mCase 4: Muestras con Cl, sin P2O5 y sin MnO: \033[0m {}'.format(len(case4)))

temp1_2 = temp1[temp1.Cl == -1]
temp1_2_1 = temp1_2[temp1_2.P2O5 != -1]
temp1_2_1_1 = temp1_2_1[temp1_2_1.MnO != -1];temp1_2_1_2 = temp1_2_1[temp1_2_1.MnO == -1]
case5 = temp1_2_1_1.replace(-1,np.nan); case6 = temp1_2_1_2.replace(-1,np.nan)
print('\033[1mCase 5: Muestras sin Cl, con P2O5 y con MnO: \033[0m {}'.format(len(case5)))
print('\033[1mCase 6: Muestras sin Cl, con P2O5 y sin MnO: \033[0m {}'.format(len(case6)))

temp1_2_2 = temp1_2[temp1_2.P2O5 == -1]
temp1_2_2_1 = temp1_2_2[temp1_2_2.MnO != -1];temp1_2_2_2 = temp1_2_2[temp1_2_2.MnO == -1]
case7 = temp1_2_2_1.replace(-1,np.nan); case8 = temp1_2_2_2.replace(-1,np.nan)
print('\033[1mCase 7: Muestras sin Cl, sin P2O5 y con MnO: \033[0m {}'.format(len(case7)))
print('\033[1mCase 8: Muestras sin Cl, P2O5 sin MnO: \033[0m {}'.format(len(case8)))

#fig_copy.savefig('../Plots/MissingnessMajorElements.png')
Data_raw_and_normalized_major = Data_raw_and_normalized_major.replace(-1,np.nan)

Case 1: Muestras con Cl, P2O5 y MnO:  2995
From samples in Case 1, 849 have corresponding geostandard information analyzed along with the samples
Case 2: Muestras con Cl, con P2O5 y sin MnO:  0
Case 3: Muestras con Cl sin P2O5 y con MnO:  1651
Case 4: Muestras con Cl, sin P2O5 y sin MnO:  0
Case 5: Muestras sin Cl, con P2O5 y con MnO:  7353
Case 6: Muestras sin Cl, con P2O5 y sin MnO:  3
Case 7: Muestras sin Cl, sin P2O5 y con MnO:  498
Case 8: Muestras sin Cl, P2O5 sin MnO:  525


Check values of P2O5, Cl and MnO

In [5]:
print('P2O5 min: {}, max: {}, mean: {}, std:{}'.format(Data_raw_and_normalized_major.P2O5.astype(float).min(),Data_raw_and_normalized_major.P2O5.astype(float).max(),Data_raw_and_normalized_major.P2O5.astype(float).mean(),Data_raw_and_normalized_major.P2O5.astype(float).std()))
print('Cl min: {}, max: {}, mean: {}, std:{}'.format(Data_raw_and_normalized_major.Cl.min(),Data_raw_and_normalized_major.Cl.max(),Data_raw_and_normalized_major.Cl.mean(),Data_raw_and_normalized_major.Cl.std()))
print('MnO min: {}, max: {}, mean: {}, std:{}'.format(Data_raw_and_normalized_major.MnO.astype(float).min(),Data_raw_and_normalized_major.MnO.astype(float).max(),Data_raw_and_normalized_major.MnO.astype(float).mean(),Data_raw_and_normalized_major.MnO.astype(float).std()))

P2O5 min: 0.0, max: 1.783379878, mean: 0.2273731760764243, std:0.14142881750842304
Cl min: 0.001000096, max: 0.61, mean: 0.14610801162656095, std:0.06135858112092851
MnO min: -0.04, max: 4.4, mean: 0.1538741186884202, std:0.15546890392177307


In [9]:
# compare chronology
from plotly.offline import plot, iplot, init_notebook_mode
import plotly.io as pio
#pio.renderers.default = "notebook"

import plotly.graph_objects as go
import plotly.express as px
import sys
sys.path.insert(1, '../Scripts')
from funciones import simbologia, color, Colores

volcanoes_by_latitude = pd.read_excel("../Scripts/VolcanesChile.xlsx")
volcanoes_by_latitude = volcanoes_by_latitude[(volcanoes_by_latitude.Activity!='Off')&
                                              (volcanoes_by_latitude.Latitud<-38.68)][['Volcan','Latitud']]
case1.sort_values('Volcano', inplace=True)

fig = px.scatter(case1, 'SiO2', 'K2O' , color = 'Volcano',
                 color_discrete_map = Colores(case1['Volcano'],type="volcano"), 
                 hover_data = ['SampleID','SampleObservationID',
                               'Volcano','Event','Authors',
                               'AnalyzedMaterial','Flag'],
                 labels = volcanoes_by_latitude['Volcan'],
                 width=800, height=500)

fig.show()
fig.write_html("../docs/CheckNormalizations_fig1.html", include_plotlyjs='cdn')# *to use the cdn option, internet conection is required

Check how much the normalized values of each element changes when renormalizing the data by a new 
Total = SiO2 + TiO2 + Al2O3 + MgO + CaO + Na2O + K2O (i.e. without MnO, P2O5 or CL)

Only data where all major elements have been analyzed (corresponding to 2995 Sample Observations from Case 1) are tested.

In [7]:
case1_renormalizado = case1.copy()
case1_renormalizado = case1_renormalizado.reset_index()
case1 = case1.reset_index()
case1_renormalizado['MnO'] = case1_renormalizado['MnO'].replace('-',-1).astype(float)
case1_renormalizado['P2O5'] = case1_renormalizado['P2O5'].replace('-',-1).astype(float)
case1_renormalizado['Cl'] = case1_renormalizado['Cl'].replace('-',-1).astype(float)

#Defining some variables which we will plot later to understand the variability of the re normalized data 
case1_renormalizado['MnO + P2O5 + Cl'] = 'default'
case1_renormalizado['Analytical Total without LOI'] = 'default'

for i in range(0,len(case1_renormalizado.Total)):
    sum_ = np.nansum([case1_renormalizado.SiO2[i],
                      case1_renormalizado.TiO2[i],
                      case1_renormalizado.Al2O3[i],
                      case1_renormalizado['FeOT'][i],
                      case1_renormalizado.MgO[i],
                      case1_renormalizado.CaO[i],
                      case1_renormalizado.Na2O[i],
                      case1_renormalizado.K2O[i]])
    
    case1_renormalizado.loc[i,'MnO + P2O5 + Cl'] = np.nansum([case1_renormalizado.MnO[i],
                                                              case1_renormalizado.Cl[i],
                                                              case1_renormalizado.P2O5[i]])
    case1_renormalizado.loc[i,'Analytical Total without LOI'] = np.nansum([case1_renormalizado.Total[i],
                                                                           -case1_renormalizado.LOI[i]])
    
    for elemento in ['SiO2','TiO2','Al2O3','FeOT','MgO','CaO','Na2O','K2O']:
        case1_renormalizado.loc[i,elemento] = case1_renormalizado[elemento][i]*100/sum_

In [8]:
#load MeasurementRuns file to now the secondary standards measured along with the samples
Data_standards = pd.read_excel('../Data/MeasurementRuns.xlsx')
Data_standards.loc[:,'MeasurementRun'] = Data_standards.loc[:,'MeasurementRun'].astype('str')
case1_renormalizado.loc[:,'MeasurementRun'] = case1_renormalizado.loc[:,'MeasurementRun'].astype('str')

#load the Secondary standards reference values
Standards_ref = pd.read_excel('../Data/Standards_Reference.xlsx')
Standards_ref = Standards_ref[Standards_ref.error.isin(['95%CL','SD'])].copy()

In [9]:
#get which secondary standards where analyzed along with the samples
stds =  Data_standards[Data_standards.MeasurementRun.isin(case1_renormalizado.MeasurementRun.unique())].StandardID.unique()
Standards_ref_case1 = Standards_ref[Standards_ref.StandardID.isin(stds)]
print(Standards_ref_case1.StandardID.unique())

['GOR132-G' 'GOR128-G' 'KL2-G' 'ML3B-G' 'StHs6/80-G' 'ATHO-G']


In [10]:
#obtaining the difference between the original normalized values (with Totals taking into consideration different elements)
#and the re-normalized values
case1_renormalizado['ΔSiO2'] = np.subtract(case1_renormalizado['SiO2'],case1['SiO2_normalized'])
case1_renormalizado['ΔTiO2'] = np.subtract(case1_renormalizado['TiO2'],case1['TiO2_normalized'])
case1_renormalizado['ΔAl2O3'] = np.subtract(case1_renormalizado['Al2O3'],case1['Al2O3_normalized'])
case1_renormalizado['ΔFeOT'] = np.subtract(case1_renormalizado['FeOT'],case1['FeOT_normalized'])
case1_renormalizado['ΔMgO'] = np.subtract(case1_renormalizado['MgO'],case1['MgO_normalized'])
case1_renormalizado['ΔCaO'] = np.subtract(case1_renormalizado['CaO'],case1['CaO_normalized'])
case1_renormalizado['ΔNa2O'] = np.subtract(case1_renormalizado['Na2O'],case1['Na2O_normalized'])
case1_renormalizado['ΔK2O'] = np.subtract(case1_renormalizado['K2O'],case1['K2O_normalized'])

In [ ]:
import plotly.express as px
import plotly.graph_objects as go

elementos = ['SiO2','TiO2','Al2O3','FeOT','MgO','CaO','Na2O','K2O']

for elemento in elementos: #['SiO2','Al2O3']:
    plot1 = go.Scatter(x=case1_renormalizado[elemento+'_normalized'],
               y=case1_renormalizado['Δ'+elemento],
               mode = 'markers',
               hovertext = 'Sample Observation ID: ' + case1_renormalizado.SampleObservationID.astype('str') + ' '+ 
                       case1_renormalizado.Authors.astype('str') ,
               marker = dict(color=case1_renormalizado['MnO + P2O5 + Cl'],showscale = True),
               name = "MnO + P2O5 + Cl")

    plot2 = go.Scatter(x = Standards_ref_case1[elemento],
                  y = Standards_ref_case1['σ_'+elemento],
                  text = Standards_ref_case1.StandardID ,
                  hovertext = 'Secondary standard: ' + Standards_ref_case1.StandardID  , 
                  mode = 'markers+text',
                  textposition = 'top center' ,
                  showlegend = False ,
                  line = {'color': "#3b3b3b"})

    Data = [plot1,plot2]
    layout = go.Layout(xaxis=dict(title=elemento), yaxis=dict(title='Δ'+elemento),
                      legend={'y':1.1, 'x':0.98}, width=800, height=500)# title='Re-normalization variation',
    
    fig = go.Figure(data=Data,layout=layout)
    #fig.show("png", engine='kaleido')

    fig.write_image('../Plots/temp.png')

To see by each MeasurementRun the relationship between the difference in the normalized values and the 95%CL interval of the analyzed secondary standards:

In [ ]:
elementos = ['SiO2','TiO2','Al2O3','FeOT','MgO','CaO','Na2O','K2O']
case1.MeasurementRun = case1.MeasurementRun.replace(np.nan,-1)

for ms in case1.MeasurementRun.unique():
    print('\033[1mMeasurement Run \033[0m {}'.format(ms))
    temp = case1_renormalizado[case1_renormalizado.MeasurementRun==ms]
    #print(temp)
    std = Data_standards[Data_standards.MeasurementRun==ms].StandardID.unique()
    temp_std = Standards_ref_case1[Standards_ref_case1.Standard.isin(std)]
    print('\033[1mAnalyzed secondary standards: \033[0m {}'.format(std))
    if any(item in std.tolist() for item in ['CFA-47','NIST2711','NCS DC70009 (GBW07241)','GBW 07113','Obsidian Astimex',
                         'JR-1','NIST694','Lipari','NCS DC70014','VG A-99','VG-2','VG-568']):
    #if std.tolist() in ['CFA-47','NIST2711','NCS DC70009 (GBW07241)','GBW 07113','Obsidian Astimex',
    #                     'JR-1','NIST694','Lipari','NCS DC70014','VG A-99','VG-2','VG-568']:
        print('No certified values available for secondary standard')
    if ms != -1:
        for elemento in ['SiO2']: #elementos:
            plot1 = go.Scatter(x=temp[elemento],
               y=temp['Δ'+elemento],
               mode = 'markers',
               hovertext = 'Sample Observation ID: ' + temp.SampleObservationID.astype('str') + ' '+ 
                       temp.Authors.astype('str') ,
               marker = dict(color=temp['MnO + P2O5 + Cl'],showscale = True),
               name = "MnO + P2O5 + Cl")

            plot2 = go.Scatter(x = temp_std[elemento],
                  y = temp_std['σ_'+elemento],
                  text = temp_std.Standard ,
                  hovertext = 'Secondary standard: ' + temp_std.Standard  , 
                  mode = 'markers+text',
                  textposition = 'top center' ,
                  showlegend = False ,
                  line = {'color': "#3b3b3b"})

            Data = [plot1,plot2]
            layout = go.Layout(xaxis=dict(title=elemento), yaxis=dict(title='Δ'+elemento),
                      legend={'y':1.1, 'x':0.98}, width=800, height=500)# title='Re-normalization variation',
    
            fig = go.Figure(data=Data,layout=layout)
            fig.show('png')
    #fig.write_image('../Plots/Δ'+elemento+'.pdf')